### DISCUSSION TITLE
Recursive generic types

### DISCUSSION THREAD

**Discussion by vioel** [microsoft/pyright#4696] [Q&A]
Hi!

I am trying to implement a function that converts a nested iterator to a nested list (or any other nested sequence); consuming all iterators in the process. The function serves its purpose, but the type narrowing does not work as expected.

```python
_T1 = TypeVar('_T1')

MaybeIterator = Union[
    _T1, Iterator['MaybeIterator[_T1]']
]
MaybeList = Union[
    _T1, List['MaybeList[_T1]']
]

def consume(
    iterators: Iterator[
        MaybeIterator[_T1]
    ]
):
    def _consume(
        iterators: Iterator[
            MaybeIterator[_T1]
        ]
    ):
        values: List[MaybeList[_T1]]
        values = list()

        for iterator in iterators:
            values.append(
                _consume(iterators=iterator)
                    if isinstance(iterator, Iterator) else iterator
            )

        return values

    return _consume(
        iterators=iterators
    )
```

I would expect the 'iterator' variable to be narrowed to 'MaybeIterator[_T1]' but pyright (pylance) give the following error:

> (variable) iterator: Iterator[Unknown]* | Iterator[MaybeIterator]
> Argument type is partially unknown
>   Argument corresponds to parameter "iterators" in function "_consume"
>   Argument type is "Iterator[Unknown]* | Iterator[_T1@MaybeIterator | Iterator[MaybeIterator]]"

Is there any way for pyright to not infer the type as 'Iterator[Unknown]* | Iterator[MaybeIterator]' ?

**Comment by erictraut** [comment]
It looks like you have "strict" type checking enabled. If you don't want to see errors about `Unknown` (implicit `Any`) types, you can use the default "basic" type checking mode.

In your code sample, when you perform the check `isInstance(iterator, Iterator)`, the type checker must assume that the type variable `_T1` might potentially be an iterator. That generates the `Iterator[Unknown]*` subtype.

I don't see a way to eliminate the `Unknown` in this case, so this may be a situation where a `cast` or `# type: ignore` is unavoidable.

There is a new type feature that I proposed (and prototyped in pyright) that would solve your problem. It's an extension of the `TypeGuard` capability introduced in PEP 647. This extension is called `StrictTypeGuard`. Here's how it would work in your code:

```python
from typing import Any, Iterator, TypeVar
from typing_extensions import StrictTypeGuard

_T1 = TypeVar("_T1")
MaybeIterator = _T1 | Iterator["MaybeIterator[_T1]"]
MaybeList = _T1 | list["MaybeList[_T1]"]

def consume(iterators: Iterator[MaybeIterator[_T1]]):
    def _is_iterator(val: object) -> StrictTypeGuard[Iterator[Any]]:
        return isinstance(val, Iterator)

    def _consume(iterators: Iterator[MaybeIterator[_T1]]):
        values: list[MaybeList[_T1]] = []

        for iterator in iterators:
            values.append(
                _consume(iterators=iterator) if _is_iterator(iterator) else iterator
            )

        return values

    return _consume(iterators=iterators)
```

While pyright supports `StrictTypeGuard` for prototyping purposes, it is not yet formally spec'ed or ratified, and it's not supported at runtime.
